module/interface.py

```diff
@@ -130,93 +130,73 @@ def setup(self, P=None, q=None, A=None, l=None, u=None, **settings):
+            elif len(l) != m:
                 raise ValueError("l must have length m")
-
-            # Convert values to OSQP_INFTY
+            # Convert values to -OSQP_INFTY
             l = np.maximum(l, -self._model.constant('OSQP_INFTY'))
-
-            if u is None:
-                self._model.update_lower_bound(l)
-
-        # Update upper bound
         if u is not None:
-            if len(u) != m:
+            if not isinstance(u, np.ndarray):
+                raise TypeError("u must be numpy.ndarray, not %s" % type(u).__name__)
+            elif len(u) != m:
                 raise ValueError("u must have length m")
-
             # Convert values to OSQP_INFTY
             u = np.minimum(u, self._model.constant('OSQP_INFTY'))
+        if Ax is None:
+            if len(Ax_idx) > 0:
+                raise ValueError("Vector Ax has not been specified")
+        else:
+            if len(Ax_idx) > 0 and len(Ax) != len(Ax_idx):
+                raise ValueError("Ax and Ax_idx must have the same lengths")
+        if Px is None:
+            if len(Px_idx) > 0:
+                raise ValueError("Vector Px has not been specified")
+        else:
+            if len(Px_idx) > 0 and len(Px) != len(Px_idx):
+                raise ValueError("Px and Px_idx must have the same lengths")
+
+        # update linear cost
+        if q is not None:
+            self._model.update_lin_cost(q)
 
-            if l is None:
-                self._model.update_upper_bound(u)
+        # update lower bound
+        if l is not None and u is None:
+            self._model.update_lower_bound(l)
 
-        # Update bounds
+        # update upper bound
+        if u is not None and l is None:
+            self._model.update_upper_bound(u)
+
+        # update bounds
         if l is not None and u is not None:
             self._model.update_bounds(l, u)
 
-        # Update matrix P
-        if Px is not None:
-            if Px_idx is not None and len(Px) != len(Px_idx):
-                raise ValueError("Px and Px_idx must have same length")
-            if Ax is None:
-                self._model.update_P(Px, Px_idx, len(Px))
-
-        # Update matrix A
-        if Ax is not None:
-            if Ax_idx is not None and len(Ax) != len(Ax_idx):
-                raise ValueError("Ax and Ax_idx must have same length")
-            if Px is None:
-                self._model.update_A(Ax, Ax_idx, len(Ax))
-
-        # Update matrices P and A
+        # update matrix P
+        if Px is not None and Ax is None:
+            self._model.update_P(Px, Px_idx, len(Px))
+
+        # update matrix A
+        if Ax is not None and Px is None:
+            self._model.update_A(Ax, Ax_idx, len(Ax))
+
+        # update matrices P and A
         if Px is not None and Ax is not None:
             self._model.update_P_A(Px, Px_idx, len(Px), Ax, Ax_idx, len(Ax))
 
-        if q is None and \
```

Comment:
It seems that if the user passes some stupid kwarg the function does not do anything nor complain. This might be an issue because because the user cannot see that.

Reply:
The function will actually through an error. Notice that we don't use kwargs any more. The function now takes only `q`, `l`, `u`, `Px`, `Px_idx`, `Ax`, `Ax_idx` as arguments.

Reply:
I guess we should only prevent a user from calling `update` with no arguments.

Reply:
Done.